pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,554 @@
+    match res {
+        Ok(outcome) => {
+            debug!(?outcome, "disk_usage_eviction_iteration finished");
+            match outcome {
+                IterationOutcome::NoPressure | IterationOutcome::Cancelled => {
+                    // nothing to do, select statement below will handle things
+                }
+                IterationOutcome::Finished(outcome) => {
+                    if outcome.after_actual.has_pressure() {
+                        // Don't bother doing an out-of-order iteration here now.
+                        // In practice, the task period is set to a value in the tens-of-seconds range,
+                        // which will cause another iteration to happen soon enough.
+                        // TODO: deltas between the three different usages would be helpful,
+                        // consider MiB, GiB, TiB
+                        // TODO: increase a counter when this happen => alert
+                        warn!(?outcome, "disk usage still high");
+                    } else {
+                        info!(?outcome, "disk usage pressure relieved");
+                    }
+                }
+            }
+        }
+        Err(e) => {
+            error!("disk_usage_eviction_iteration failed: {:#}", e);
+        }
+    }
+
+    let sleep_until = start + task_config.period;
+    tokio::select! {
+        _ = tokio::time::sleep_until(sleep_until) => ControlFlow::Continue(()),
+        _ = cancel.cancelled() => {
+            info!("shutting down");
+            ControlFlow::Break(())
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(clippy::large_enum_variant)]
+enum IterationOutcome {
+    NoPressure,
+    Cancelled,
+    Finished(IterationOutcomeFinished),
+}
+
+// The `#[allow(dead_code)]` is to suppress warnings about only the Debug impl reading these fields.
+// We use the Debug impl for logging, so, it's allright.
+#[allow(dead_code)]
+#[derive(Debug)]
+struct IterationOutcomeFinished {
+    before: UsageWithPressure,
+    after_actual: UsageWithPressure,
+    after_expected: Usage,
+
+    // Ok(false) a bit problematic
+    evictions_failed: LayerCount,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Default)]
+struct LayerCount {
+    file_sizes: u64,
+    count: usize,
+}
+
+async fn disk_usage_eviction_task_iteration_impl(
+    task_config: &DiskUsageEvictionTaskConfig,
+    storage: &GenericRemoteStorage,
+    tenants_dir_fd: &mut SyncWrapper<Dir>,
+    cancel: &CancellationToken,
+) -> anyhow::Result<IterationOutcome> {
+    let usage_pre = filesystem_level_usage::get(tenants_dir_fd)?.with_pressure(task_config);
+    // planned post-eviction usage
+    let mut usage_expected = usage_pre.usage.clone();
+    // achieved post-eviction usage according to internal accounting
+    let mut usage_assumed = usage_pre.usage.clone();
+    // actual usage read after batched evictions
+    let usage_post;
```

Comment:
Clippy dislikes this. However, we'd prefer to have them in the same place, or at least yesterday it seemed like a good idea.